vmm/src/vm.rs

```diff
@@ -2131,6 +2142,15 @@ impl Vm {
     pub fn restore(&mut self) -> Result<()> {
         event!("vm", "restoring");
 
+        // Note: Always call this function before invoking start restored vcpus.
+        // Otherwise guest would fail to boot because we haven't created the
+        // userspace mappings to update the hypervisor about the memory mappings.
+        self.memory_manager
```

Comment:
@michael2012z can we call 

```
        device_manager
            .lock()
            .unwrap()
            .create_devices(serial_pty, console_pty, console_resize_pipe)
            .map_err(Error::DeviceManager)?;
```

right before the call to `allocate_address_space()`?

Reply:
No, `create_devices()` need to be called after `allocate_address_space()`.

Reply:
Yeah sorry I had the reverse order in that statement. I have attempted to do that in this commit: https://github.com/cloud-hypervisor/cloud-hypervisor/pull/5169/commits/f5ba36b80c52ef9a3c7de3808b82e9446973cf7a

Reply:
Sadly, that won't work because the ACPI table generation do need those devices to be present.

Reply:
And I cannot delay the ACPI table generation because that would mean the contents of the guest memory is modified after I have created the userspace mappings. So, there is no easy way to unify this path on x86 and aarch64. @liuw @rbradford @likebreath any thoughts if we split this code path for x86 and aarch64?

Reply:
I think this is fine.